vm/devices/storage/disk_nvme/nvme_driver/src/queue_pair.rs

```diff
@@ -280,43 +303,100 @@ impl Issuer {
             .probe_gpns(mem.gpns())
             .map_err(RequestError::Memory)?;
 
-        let prp = if mem
-            .gpns()
-            .iter()
-            .all(|&gpn| guest_memory.iova(gpn * PAGE_SIZE64).is_some())
+        // TODO: add check if the memeory if VA-backed.
+        let is_va_backed = false;
+        let (prp, is_memory_pinned) = if !is_va_backed
+            && mem
+                .gpns()
+                .iter()
+                .all(|&gpn| guest_memory.iova(gpn * PAGE_SIZE64).is_some())
         {
             // Guest memory is available to the device, so issue the IO directly.
-            self.make_prp(
-                mem.offset() as u64,
-                mem.gpns()
-                    .iter()
-                    .map(|&gpn| guest_memory.iova(gpn * PAGE_SIZE64).unwrap()),
+            (
+                self.make_prp(
+                    mem.offset() as u64,
+                    mem.gpns()
+                        .iter()
+                        .map(|&gpn| guest_memory.iova(gpn * PAGE_SIZE64).unwrap()),
+                )
+                .await,
+                false,
             )
-            .await
         } else {
-            tracing::debug!(opcode = opcode.0, size = mem.len(), "double buffering");
-
-            // Guest memory is not accessible by the device. Double buffer
-            // through an allocation.
-            let double_buffer_pages = double_buffer_pages.insert(
-                self.alloc
-                    .alloc_bytes(mem.len())
-                    .await
-                    .ok_or(RequestError::TooLarge)?,
-            );
+            // Guest memory is not accessible by the device.
+            // If guest memory is VA-backed, and IO size exceeds threshold, pin the memory.
+            let mut prp_result = None;
+            let mut is_pinned = false;
+            if let Some(io_threshold) = self.io_threshold {
+                if is_va_backed && self.partition.is_some() && mem.len() as u32 > io_threshold {
```

Comment:
In addition, let's factor out the check for io size > io_threshold as well. The routine should either pin or buffer. It can take in the threshold as one config. It should also figure out if the memory is VA backed.

Okay, I still think I'm talking about what I think John will scaffold. But I want to give the feedback here as well: this code is challenging to follow, so I'm making point suggestions on how to factor the code so that (IMO) it'll be easier to read and maintain in the future.

Reply:
Yeah, I believe the logic can be moved from the nvme_driver to the common DMA APIs once John's design is done. We might not need the io_threshold and can always buffer first if there's enough space. However, considering performance, adding the io_threshold could benefit larger IOs and provide flexibility for performance tuning later.